### PR TITLE
Add the tenant into the short URL once the short URL is resolved

### PR DESCRIPTION
**Pull request by leanneeliatra**
### Description
This is a duplicate PR (closed) to https://github.com/opensearch-project/security-dashboards-plugin/pull/1443,
Copying the changes from a branch branched from main allowed me a cleaner & safer way to perform similar rebase.
__________________________
When the tenancy is being changed after copying a short URL, if the user then pastes this short URL into a new tab, the user will be redirected to the current selected tenant, not the tenant as specified in the copied short URL. 
We need to retain this information in the URL and redirect so the user is brought to the correct link, with the correct user.

### Category
[Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation]

### Why these changes are required?
To enable sharing of a short URL with the tenant as specified in the URL remaining selected as the active tenant.

### What is the old behavior before changes and new behavior after changes?
The old behaviour is: 
1. A short URL is copied (tenant is specified in the URL)
2. The tenant is changed
3. The link is loaded in a new tab
4. The tenant remains as is specified in 2, not as we would want/expect it to be as per the URL 
The new behaviour is: 
1. A short URL is copied (tenant is specified in the URL)
2. The tenant is changed
3. The link is loaded in a new tab
4. The link is loaded with tenant info retained.

### The goal of the piece of code 
* [Solve issue 1203](https://github.com/opensearch-project/security-dashboards-plugin/issues/1203) which led to the user being directed to the incorrect tenant. This was happening in the case the short URL was copied, then the tenant was changed & the URL was loaded in a new tab.The outcome was the tenant specified in the short URL was not maintained on page load.

###  How it is done 
* The short URL when copied and pasted has the security_tenant, however, when the URL is being rendered and parsed, this information was not retained in the resolved short URL.
* Upon page reload, the security tenant was fetched again from the localstorage, which was the switched tenant (not the tenant originally specified in the short URL.)

### Why it is done this way
* Therefore in the case of multitenancy, when parsing the short URL, we inject the security tenant into the short URL, adding it as a parameter and retaining this information then in the resolved URL. Allowing the page to correctly load the page, with the correct tenant that was specified & copied in the short URL.
* The addTenantParameterToResolvedShortLink controls this functionality.

#### Other attempted solutions:
The first solution implemented to fix this issue, was to wipe the local storage to solve this issue.
However, once it was done, I realised the localstorage value was needed elsewhere & could not be wiped as intended. 
New solutions were considered, perhaps using a cookie, which came with it's own challenges to change how localstorage is currently being used.
In the end, the solution that maintained the localstorage (as it was needed in the code for other purposes) and solved the issue at hand (couldn't change the tenant after copying the short URL), was to, if the tenant was changed in multitenancy mode, to make sure the tenant information remained in the short URL (it was there before page load but is removed from the resolved URL, we ensured the tenant information was retained in the resolved URL) and also, remove any specifics of the URL (so the user wouldn't be brought to a dashboard that doesn't exist in that tenant and get an error message). 

### Issues Resolved
[[BUG] Shortlinks to other Tenants do not work #1203](https://github.com/opensearch-project/security-dashboards-plugin/issues/1203)

### Testing
Submitting the PR for an initial review of changes, tests will follow.

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/plugin.ts
Comment (leanneeliatra) [reply server/plugin.ts]: I have moved the main body of information from the comments into the PR itself, and left much smaller comments in the code, thanks @scrawfor99.

File: server/plugin.ts
Comment (leanneeliatra) [reply server/plugin.ts:133]: I have just added the tests for this code but I will come back to review in light of multi tenancy tests too (once I have tests for 1204 in), thanks @scrawfor99

Comment (leanneeliatra) [reply server/plugin.ts:133]: The tests added cover these cases, thanks @scrawfor99

File: server/multitenancy/tenant_resolver.ts
Comment (DarshitChanpura) [inline server/multitenancy/tenant_resolver.ts:223]: why check for this condition?

File: server/plugin.ts
Comment (DarshitChanpura) [inline server/plugin.ts:130]: nit: can skip this comment line

File: server/multitenancy/tenant_resolver.ts
Comment (DarshitChanpura) [inline server/multitenancy/tenant_resolver.ts:217]:
```suggestion
if (parts.query.security_tenant === undefined)
```

File: server/plugin.ts
Comment (leanneeliatra) [reply server/plugin.ts:130]: perfect thanks I have removed this.

File: server/multitenancy/tenant_resolver.ts
Comment (leanneeliatra) [reply server/multitenancy/tenant_resolver.ts]: Thanks Darshit I see now there is no need for it as it's checked further up in the function.

I've amended the code.

**Review by DarshitChanpura — COMMENTED** [review]
LGTM. Waiting on CI before approval. Thank you for being on top of addressing comments

**Review by peternied — COMMENTED** [review]
Can we add or modify an integration test for this scenario?  Thanks for the unit tests; testing this in a browser seems like a requirement to know if this issue is fixed.

**Comment by leanneeliatra** [comment]
> Details

Hi @DarshitChanpura, some of the tests seem to be failing. Do you have any guidance on what I can do to solve this please?

**Comment by peternied** [comment]
@scrawfor99 Could you look into the failures, seems like there is an issue with the ubuntu build.  Might be from OpenSearch, it might be from security?

```
java.lang.Exception: java.lang.NoClassDefFoundError: org/opensearch/common/lease/Releasable
	at org.opensearch.transport.netty4.Netty4MessageChannelHandler.exceptionCaught(Netty4MessageChannelHandler.java:106) [transport-netty4-client-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
...
```

**Comment by nibix** [comment]
@peternied 

> Can we add or modify an integration test for this scenario? Thanks for the unit tests; testing this in a browser seems like a requirement to know if this issue is fixed.

Sure. 

I would like to check just one thing:

There is already a test suite at https://github.com/opensearch-project/security-dashboards-plugin/blob/e628c81b4589b63264158a971f95a7b5a1e5e4a1/test/jest_integration/multi_tenancy.test.ts which we could extend.

However, it seems that the preferred way of writing tests is now adding these to https://github.com/opensearch-project/opensearch-dashboards-functional-test  according to https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/main/DEVELOPER_GUIDE.md#tests-for-opensearch-dashboard-plugins

What do you think?

**Comment by peternied** [comment]
I'd bow to the wisdom of anyone with more experience in these different frameworks, @RyanL1997 @DarshitChanpura do you have thoughts for testing this scenario?

My off the cuff suggestion would be to write them with jest since its in this repository.

**Comment by leanneeliatra** [comment]
> I'd bow to the wisdom of anyone with more experience in these different frameworks, @RyanL1997 @DarshitChanpura do you have thoughts for testing this scenario?
> 
> My off the cuff suggestion would be to write them with jest since its in this repository.

Hi @DarshitChanpura & @RyanL1997, to follow on from this we had a further look into the testing and there are some multitenancy tests that have been added recently to the external functional test repo:
https://github.com/opensearch-project/opensearch-dashboards-functional-test/tree/main/cypress/integration/plugins/security-dashboards-plugin

These tests then seem to be triggered through github actions in the plugin repo:
https://github.com/opensearch-project/security-dashboards-plugin/blob/main/.github/workflows/cypress-test-tenancy-disabled.yml#L76

Those tests are cypress test, it seems that could be a good path forward? Thanks.

File: server/multitenancy/tenant_resolver.ts
Comment (cwperks) [inline server/multitenancy/tenant_resolver.ts:218]: Is the header always `securitytenant`? I see code in `tenant_resolver.ts` that also shows `securityTenant_`: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/multitenancy/tenant_resolver.ts#L72-L75

**Comment by leanneeliatra** [comment]
> Hey @leanneeliatra I'd suggest writing a unit test in jest and an end-to-end test in functional test repo.

Thanks @DarshitChanpura, I'll do that.

File: server/multitenancy/tenant_resolver.ts
Comment (cwperks) [reply server/multitenancy/tenant_resolver.ts:218]: The only references I see to `securityTenant_` are for url params so I think this is fine.

**Review by DarshitChanpura — APPROVED** [review]

**Pull request merged**
